WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt

```diff
@@ -0,0 +1,112 @@
+import org.wordpress.android.viewmodel.posts.PostListItemType.EndListIndicatorItem
+import org.wordpress.android.viewmodel.posts.PostListItemType.LoadingItem
+import org.wordpress.android.viewmodel.posts.PostListItemType.PostListItemUiState
+
+sealed class PostListItemIdentifier {
+    data class LocalPostId(val id: LocalId) : PostListItemIdentifier()
+    data class RemotePostId(val id: RemoteId) : PostListItemIdentifier()
+    object EndListIndicatorIdentifier : PostListItemIdentifier()
+}
+
+class PostListItemDataSource(
+    private val dispatcher: Dispatcher,
+    private val postStore: PostStore,
+    private val performGetItemIdsToHide: ((PostListDescriptor) -> Pair<LocalPostId, RemotePostId>?),
+    private val transform: ((PostModel) -> PostListItemUiState)
+) : ListItemDataSourceInterface<PostListDescriptor, PostListItemIdentifier, PostListItemType> {
+    override fun fetchList(listDescriptor: PostListDescriptor, offset: Long) {
+        val fetchPostListPayload = FetchPostListPayload(listDescriptor, offset)
+        dispatcher.dispatch(PostActionBuilder.newFetchPostListAction(fetchPostListPayload))
+    }
+
+    override fun getItemIdentifiers(
+        listDescriptor: PostListDescriptor,
+        remoteItemIds: List<RemoteId>,
+        isListFullyFetched: Boolean
+    ): List<PostListItemIdentifier> {
+        val localItems = postStore.getLocalPostIdsForDescriptor(listDescriptor)
+                .map { LocalPostId(id = it) }
+        val remoteItems = remoteItemIds.map { RemotePostId(id = it) }
+        val actualItems = localItems + remoteItems
+
+        val endListItem = if (isListFullyFetched && actualItems.isNotEmpty()) {
+            listOf(EndListIndicatorIdentifier)
+        } else emptyList()
+        val itemsToHide = getItemIdsToHide(listDescriptor)
+        // We only want to show the end list indicator if the list is fully fetched and it's not empty
+        return (actualItems + endListItem).filter { !itemsToHide.contains(it) }
+    }
+
+    override fun getItemsAndFetchIfNecessary(
+        listDescriptor: PostListDescriptor,
+        itemIdentifiers: List<PostListItemIdentifier>
+    ): List<PostListItemType> {
+        val localOrRemoteIds = itemIdentifiers.mapNotNull { identifier ->
+            when (identifier) {
+                is LocalPostId -> identifier.id
+                is RemotePostId -> identifier.id
+                // We are creating a list of local and remote ids, so other type of identifiers don't matter
+                EndListIndicatorIdentifier -> null
+            }
+        }
+
+        val postList = postStore.getPostsByLocalOrRemotePostIds(localOrRemoteIds, listDescriptor.site)
+
+        // Convert the post list into 2 maps with local and remote ids as keys
+        val localPostMap = postList.associateBy { LocalId(it.id) }
+        val remotePostMap = postList.filter { it.remotePostId != 0L }.associateBy { RemoteId(it.remotePostId) }
```

Comment:
Nitpick: Can we replace it.remotePostId != 0L with !it.isLocalDraft or is there any other scenario when the remotePostId == 0?

Reply:
I am not sure if those two properties are connected or not, but if they are in the FluxC side, I think we'll have to assume the implementation detail here. We need the remote id to be able to fetch posts and even if they mean the same thing, I feel like this expresses the intend a bit better.